File: core/src/main/java/org/selyu/smp/core/data/Store.java
package org.selyu.smp.core.data;

import org.jetbrains.annotations.NotNull;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public interface Store<KEY, VALUE> {
    @NotNull
    CompletableFuture<Optional<VALUE>> getByKey(@NotNull KEY key);

    @NotNull
    CompletableFuture<VALUE> save(@NotNull VALUE value);

    @NotNull
    CompletableFuture<Void> delete(@NotNull VALUE value);
}

File: core/src/main/kotlin/org/selyu/smp/core/data/Store.kt
package org.selyu.smp.core.data

import java.util.*
import java.util.concurrent.CompletableFuture

interface Store<K, V> {
    fun getByKey(key: K): CompletableFuture<Optional<V>>
    fun insert(value: V): CompletableFuture<V>
    fun delete(value: V): CompletableFuture<Void>
}